### ISSUE TITLE
Disallow custom authentication class to return user = `None`

### ISSUE THREAD

**Issue by leandrodesouzadev** [encode/django-rest-framework#9574]
Hello there!
Currently it's allowed that an custom authentication class returns a tuple containing `(None, None)`. This causes the authentication to be considered successful and sets `request.user = None` and this can cause errors that are hard to track where they came from.
I know that this wrongly implemented by the `CustomAuthentication` class, the user should've raised an `AuthenticationFailed` exception instead. But this allows the user to shoot itself on the foot.

Example of a bad implemented authentication class:
```python
from rest_framework.authentication import BaseAuthentication

class MyDumbAuthentication(BaseAuthentication):
  def authenticate(self, request):
    return None, None
```

Later if you have a permission check for example, you would see the following error:
```python
from rest_framework.permissions import BasePermission

class MyPermCheck(BasePermission):
  def has_permission(self, request, view):
    return request.user.has_perm("foo.bar")

# raises AttributeError("'NoneType' object has no attribute 'has_perm'")
```

It would be nice if rest framework disallowed this totally wrong implementation.

## Checklist

- [x] Raised initially as discussion
- [x] This is not a feature request suitable for implementation outside this project. Please elaborate what it is:
  - [ ] compatibility fix for new Django/Python version ...
  - [x] other type of bug fix
  - [ ] other type of improvement that does not touch existing code or change existing behavior (e.g. wrapper for new Django field)
- [x] I have reduced the issue to the simplest possible case.

**Comment by leandrodesouzadev** [comment]
I would be happy to submit a patch to disallow this!

**Comment by krishnamadhavan** [comment]
@leandrodesouzadev All implementations of `permission_classes` from `rest_framework` uses something like below:

```
return bool(request.user and request.user.is_staff)
```

So, you always verify `request.user` before calling any methods on `request.user`. 

Does that answer your question? Happy to help otherwise.

**Comment by Natgho** [comment]
```python
        if user is None:
            raise exceptions.AuthenticationFailed(_('Invalid username/password.'))

        if not user.is_active:
            raise exceptions.AuthenticationFailed(_('User inactive or deleted.'))

        return (user, None)
```
Those who implement poorly are also responsible for carrying the burden of maintenance like it's a crime :)

The solution above will suffice. I think this ticket can be closed @browniebroke

**Comment by browniebroke** [comment]
> It would be nice if rest framework disallowed this totally wrong implementation.

Yes, it would be nice, but not sure what else we can be done at the framework level I'm afraid... Ideally I would like a very loud ImproperlyConfigured error or something, but I don't think that's feasible here. It would still have to be at runtime, probably be when we try to authenticate if the custom auth class return this special value, so not sure how useful that would be or if that could break some deployment that somehow rely on this behaviour...

>         if user is None:
>             raise exceptions.AuthenticationFailed(_('Invalid username/password.'))
> 
>         if not user.is_active:
>             raise exceptions.AuthenticationFailed(_('User inactive or deleted.'))
> 
>         return (user, None)
>
> Those who implement poorly are also responsible for carrying the burden of maintenance like it's a crime :)

Not sure what you mean, are you suggesting that this snippet should be added in user code or in DRF? I assume the former

**Comment by Natgho** [comment]
> > It would be nice if rest framework disallowed this totally wrong implementation.
> ...
> > Those who implement poorly are also responsible for carrying the burden of maintenance like it's a crime :)
> 
> Not sure what you mean, are you suggesting that this snippet should be added in user code or in DRF? I assume the former

What I mean is that the user needs to add this to their own code. Adding something like this for DRF would likely break the code of many people who already use DRF.

In summary, since my comments are specific to the user, this ticket can be closed. It's a personal issue.

**Comment by browniebroke** [comment]
Agreed. I don't really see how to make this without potentially breaking some use cases.
